### PR TITLE
Intoduce torch.testing._internal.opinfos

### PR DESCRIPTION
**Pull request by antocuni**
@mruberry this PR is still not complete but I'd like to get some feedback before continuing, to avoid doing unnecessary work.
I started to experiment with what we discussed in #55159, and in particular:

1. `op_db` is not an instance of `OpInfoDB` instead of being a simple list. This allows to `register()` `OpInfo`s from multiple places, instead of having to list them explicitly into a big long list
2. I moved `linalg.det` and `linalg.cond` in their own file, to have a feeling on how the new system looks like. A good pro is that now we can implement the `sample_inputs_*` functions in the proximity of the `OpInfo` itself, instead of having them many lines above.

Looking forward to get some feedback!

### CODE REVIEW

**Review by mruberry — COMMENTED** [review]
What do you think, @kshitij12345, @pmeier?

common_methods_invocations.py is currently 7500 lines, and we like to keep files below 10k lines. So I think we have time before we have to pursue a split. If/when we do pursue a split I agree this seems like the best approach.

Should we do this now, @antocuni? What do you see as the pros/cons? My attempt at a list:

PROS

- nicer file names (because common_methods_invocations.py is a terrible name)
- can put sample inputs next to ops, at least for some operators
- moves classes like UnaryUfuncInfo into their own files, so files have clearer purposes

CONS

- can no longer search all opinfos in one file
- sample inputs next to ops or an op after op pattern both seem valid, and some classes of operator (unary elementwise, binary elementwise, reductions) will use the same sample inputs for the entire class, so at least a few will use an op after op pattern; this makes PRO #2 above a little dubious

What other PROS/CONS should we consider, @antocuni?

**Comment by antocuni** [comment]
> What do you think, @kshitij12345, @pmeier?
> 
> common_methods_invocations.py is currently 7500 lines, and we like to keep files below 10k lines. So I think we have time before we have to pursue a split. If/when we do pursue a split I agree this seems like the best approach.

just my 2 cents here, but generally speaking 10k lines seems a LOT to me, also considering that Python source code is typically more dense than other languages. It makes it impossible to read the file from top to bottom and the only reasonable way to navigate around it is to search for a known string. Even 7.5k is already too much, at least for me.

Also, it seems that the `OpInfo` approach is really successful and we will continuing to add new features and new operators, so eventually we will reach the 10k limit anyway. So, the sooner we do the split, the less work we have to do later when we will inevitably need to do it.

> Should we do this now, @antocuni? What do you see as the pros/cons? My attempt at a list:
> 
> PROS
> 
> * nicer file names (because common_methods_invocations.py is a terrible name)

+1 for this :)

> * can put sample inputs next to ops, at least for some operators
> * moves classes like UnaryUfuncInfo into their own files, so files have clearer purposes
> 
> CONS
> 
> * can no longer search all opinfos in one file

this can be mitigated by using a sane naming convention for the files, so that one can easily `grep` into them. In this PR I started using `opinfo_*.py`, but we could even use a separate package `opinfo/*.py` (e.g., `opinfo/linalg.py`).

> * sample inputs next to ops or an op after op pattern both seem valid, and some classes of operator (unary elementwise, binary elementwise, reductions) will use the same sample inputs for the entire class, so at least a few will use an op after op pattern; this makes PRO #2 above a little dubious

yes, but there are others cases in which there are ~6000 lines between the definition of the `sample_input_*` function and the corresponding `OpInfo`. Even if the PRO #2 does not applies everywhere, it's still a big improvements when it does.


> What other PROS/CONS should we consider, @antocuni?

Extra PROS

* with this approach, we can easily combine multiple `OpInfoDB` into one. For example, right now we start from a single big `op_db` and from there we create smaller dbs by filtering out what we don't want, like `unary_ufuncs`, `spectral_funcs`, etc. We could do the opposite and have `opinfo/unary.py` to define `unary_db`, `opinfo/spectral.py` to define `spectral_func_db` and `opinfo/all.py` to be defined as `unary_db + spectral_db + other_ops_db + ...` (but we need to find a better name for `other_ops_db`)

* it makes it possible to split the logic and the code of `OpInfo`, `OpInfoDB`, `SampleInput` etc. from the actual definition of the DB itself (which will be in the `opinfo/*.py` files)

* it enables us to experiment with new/better syntax to declare `OpInfo`s, either now or in the future. See e.g. the class-based syntax which I proposed in #55159

Extra CONS

* we need to repeat a lot of `import` boilerplate at the beginning of each file

**Comment by kshitij12345** [comment]
> nicer file names (because common_methods_invocations.py is a terrible name)

+1

> sample inputs next to ops or and op after op pattern both seem valid, and some classes of operator (unary elementwise, binary elementwise, reductions) will use the same sample inputs for the entire class, so at least a few will use an op after op pattern; this makes PRO #2 above a little dubious

I agree with @antocuni that this structure gives us more flexibility going forward.

Also from my personal experience, just looking at `test_torch.py` was intimidating (as a new contributor back when it was 15K lines and over). I really like that with this approach we can split the OpInfo into coherent files.

Thanks @antocuni!

**Comment by pmeier** [comment]
Very much in favor of having multiple smaller files. What about a structure like this:

```
torch/
└── testing
    └── _internal
        └── op_info
            ├── binary.py
            ├── __init__.py
            ├── linalg.py
            ├── misc.py
            ├── reduction.py
            ├── unary.py
            └── utils.py
```

We could group every operator in the `torch` namespace into `unary`, `binary`, `reduction`, and `misc` (or split this further if we like to). For operators that have their own namespace, I would just keep it for example `linalg`. Stuff that is used by all files, for example the base `OpInfo` class could be put in `utils`. As @antocuni proposed, every file could now define its own small database which could be fused in `__init__`.

**Comment by antocuni** [comment]
> Very much in favor of having multiple smaller files. What about a structure like this:
> 
> ```
> torch/
> └── testing
>     └── _internal
>         └── op_info
>             ├── binary.py
>             ├── __init__.py
>             ├── linalg.py
>             ├── misc.py
>             ├── reduction.py
>             ├── unary.py
>             └── utils.py
> ```

+1 for this.

> We could group every operator in the `torch` namespace into `unary`, `binary`, `reduction`, and `misc` (or split this further if we like to). For operators that have their own namespace, I would just keep it for example `linalg`. Stuff that is used by all files, for example the base `OpInfo` class could be put in `utils`. As @antocuni proposed, every file could now define its own small database which could be fused in `__init__`.

Defining an `OpInfoDB` in each file and merge them together inside `__init__` has another advantage: in the current version of the branch, `import torch.testing._internal.opinfo_linalg` has the side effect of mutating `_internal.op_db` (by calling `.register()` on it) and thus has do be done at the end of the file.
By reversing the approach, we can do a much simpler and easier to understand composition. E.g. inside `op_info/__init__.py` we will have something like this:

```python
from . import unary, binary, linalg, misc # ...
op_db = OpInfoDB.merge('all', unary.op_db, binary.op_db, linalg.op_db, misc.op_db, ...)
```

No side effects, no need to do imports at the end, everything straightforward.

@mruberry what do you think?

**Comment by mruberry** [comment]
This sounds good. What about keeping our lives simpler for now and still just using a list? Lists in Python can be appended with each new OpInfo instead of declaring them in it. The central repo can just import the lists and extend or add them or whatever to create its own list.

For files, we could start with a new folder in _internal, like proposed above, called "opinfos". The files in it can mirror different test files and classes. We could start with __init__.py, core.py or opinfo.py, unary_elementwise.py, and foreach.py for example. These could define the lists "all_opinfos" (replacing the name "op_db"), define the OpInfo-related classes, a "unary_elementwise_opinfos" list, and the special "foreach_opinfos" list, which contains OpInfos that don't appear in "all_opinfos". 

That seems simple and flexible while achieving our organizational goals? It also lets us create a more specialized data structure in the future if we want to without locking us into anything today.

**Comment by antocuni** [comment]
> This sounds good. What about keeping our lives simpler for now and still just using a list? Lists in Python can be appended with each new OpInfo instead of declaring them in it. The central repo can just import the lists and extend or add them or whatever to create its own list.
> 
> For files, we could start with a new folder in _internal, like proposed above, called "opinfos". The files in it can mirror different test files and classes.

@mruberry I changed the name of this PR to reflect the this new plan and implemented it.
This PR:

- introduces `torch.testing._internal.opinfos`
- move the majority of the old `common_method_invocations.py` into multiple files in `opinfos/`
- fix the imports accordingly

This PR is going to be a conflict nightmare since `common_methods_invocations.py` is modifed very often, so to mitigate the problem for now I tried to minimize the changes. In particular, I did not rename anything, and `op_db` is still defined as a big single list inside `opinfos/db.py`.

I propose to continue the refactoring in follow up PRs, but assuming that tests are green, hopefully this one should land as quickly as possible, because the more we wait the more likely it is to become out of date.

**Comment by mruberry** [comment]
> > This sounds good. What about keeping our lives simpler for now and still just using a list? Lists in Python can be appended with each new OpInfo instead of declaring them in it. The central repo can just import the lists and extend or add them or whatever to create its own list.
> > For files, we could start with a new folder in _internal, like proposed above, called "opinfos". The files in it can mirror different test files and classes.
> 
> @mruberry I changed the name of this PR to reflect the this new plan and implemented it.
> This PR:
> 
> * introduces `torch.testing._internal.opinfos`
> * move the majority of the old `common_method_invocations.py` into multiple files in `opinfos/`
> * fix the imports accordingly
> 
> This PR is going to be a conflict nightmare since `common_methods_invocations.py` is modifed very often, so to mitigate the problem for now I tried to minimize the changes. In particular, I did not rename anything, and `op_db` is still defined as a big single list inside `opinfos/db.py`.
> 
> I propose to continue the refactoring in follow up PRs, but assuming that tests are green, hopefully this one should land as quickly as possible, because the more we wait the more likely it is to become out of date.

Cool; let me think if we can sneak this in this weekend.

**Comment by mruberry** [comment]
fyi @antocuni I haven't forgotten about this but need to find the right time to make the update, thanks for being patient

**Comment by antocuni** [comment]
> fyi @antocuni I haven't forgotten about this but need to find the right time to make the update, thanks for being patient

no problem, thank you.
I think it's not worth trying to fix the conflicts continuously, so please ping me when you are ready to do the merge, and I'll fix it

**Comment by kshitij12345** [comment]
Right now `common_method_invocations` is around 15k lines and Github has issue rendering.

Gentle ping @mruberry

**Comment by mruberry** [comment]
> Right now `common_method_invocations` is around 15k lines and Github has issue rendering.
> 
> Gentle ping @mruberry

Yeah it's become too big

**Comment by pmeier** [comment]
We should not let this go to waste. Working on the file is a pain.
